src/mscorlib/src/System/TimeZoneInfo.cs

```diff
@@ -2178,9 +2179,93 @@ static public TimeZoneInfo FindSystemTimeZoneById(string id)
+        /// </summary>
+        static public ReadOnlyCollection<TimeZoneInfo> GetSystemTimeZones()
+        {
+            // UNIXTODO: Issue #2497 use CachedData
+
+            List<TimeZoneInfo> timeZoneInfos = new List<TimeZoneInfo>();
+
+            string timeZoneDirectory = GetTimeZoneDirectory();
+            foreach (string timeZoneId in GetTimeZoneIds(timeZoneDirectory))
+            {
+                if (!string.IsNullOrEmpty(timeZoneId))
+                {
+                    try
+                    {
+                        TimeZoneInfo timeZoneInfo = FindSystemTimeZoneById(timeZoneId);
+                        timeZoneInfos.Add(timeZoneInfo);
+                    }
+                    catch (InvalidTimeZoneException) { }
+                    catch (TimeZoneNotFoundException) { }
+                }
+            }
+
+            // sort and copy the TimeZoneInfo's into a ReadOnlyCollection for the user
+            timeZoneInfos.Sort(new TimeZoneInfoComparer());
+
+            return new ReadOnlyCollection<TimeZoneInfo>(timeZoneInfos);
+        }
+
+        /// <summary>
+        /// Returns a collection of TimeZone Id values from the zone.tab file in the timeZoneDirectory.
+        /// </summary>
+        /// <remarks>
+        /// Lines that start with # are comments and are skipped.
+        /// </remarks>
+        private static IEnumerable<string> GetTimeZoneIds(string timeZoneDirectory)
+        {
+            string[] zoneTabFileLines = null;
+            try
+            {
+                zoneTabFileLines = File.ReadAllLines(Path.Combine(timeZoneDirectory, c_zoneTabFileName));
+            }
+            catch (IOException) { }
+            catch (UnauthorizedAccessException) { }
+
+            List<string> timeZoneIds = new List<string>();
+            if (zoneTabFileLines != null)
+            {
+                foreach (string zoneTabFileLine in zoneTabFileLines)
+                {
+                    if (!string.IsNullOrEmpty(zoneTabFileLine) && !zoneTabFileLine.StartsWith("#"))
+                    {
+                        // the format of the line is "country-code \t coordinates \t TimeZone Id \t comments"
+
+                        int firstTabIndex = zoneTabFileLine.IndexOf('\t');
+                        if (firstTabIndex != -1)
+                        {
+                            int secondTabIndex = zoneTabFileLine.IndexOf('\t', firstTabIndex + 1);
+                            if (secondTabIndex != -1)
+                            {
+                                string timeZoneId;
+                                int startIndex = secondTabIndex + 1;
+                                int thirdTabIndex = zoneTabFileLine.IndexOf('\t', secondTabIndex + 1);
+                                if (thirdTabIndex != -1)
+                                {
+                                    int length = thirdTabIndex - secondTabIndex - 1;
```

Comment:
my question was regarding if having 2 tabs can be expected in the file. if this true then we should skip all tabs after the first one we find and then continue the reading as usual. if it is invalid to have 2 tabs then we should skip this entry here instead of later.